solver/llbsolver/proc/index.go

```diff
@@ -0,0 +1,71 @@
+package proc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/containerd/containerd/platforms"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/frontend"
+	"github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+func ForceIndexProcessor(ctx context.Context, result *frontend.Result, s *llbsolver.Solver, j *solver.Job) (*frontend.Result, error) {
+	if len(result.Refs) > 0 {
+		return result, nil
+	}
+	if result.Ref == nil {
+		return nil, errors.New("no refs to operate on")
+	}
+
+	// try to determine platform from image config, fallback to current platform
+	p := platforms.DefaultSpec()
+	if imgConfig, ok := result.Metadata[exptypes.ExporterImageConfigKey]; ok {
+		var img ocispecs.Image
+		err := json.Unmarshal(imgConfig, &img)
+		if err != nil {
+			return nil, err
+		}
+
+		if img.OS != "" && img.Architecture != "" {
+			p = ocispecs.Platform{
+				Architecture: img.Architecture,
+				OS:           img.OS,
+				OSVersion:    img.OSVersion,
+				OSFeatures:   img.OSFeatures,
+				Variant:      img.Variant,
+			}
+		}
+	}
+	p = platforms.Normalize(p)
+	pk := platforms.Format(p)
+
+	result.Refs = map[string]solver.ResultProxy{
+		pk: result.Ref,
+	}
+	result.Ref = nil
+
+	if result.Metadata != nil {
+		for _, key := range exptypes.ExporterKeys {
+			if value, ok := result.Metadata[key]; ok {
+				result.Metadata[fmt.Sprintf("%s/%s", key, pk)] = value
+				delete(result.Metadata, key)
```

Comment:
Maybe we should only remove the keys we know about. And not convert keys like `ExporterPlatformsKey` that we know about as well.

Reply:
Done, only the keys in `exptypes.ExporterPlatformKeys` will get modified now.